remove special case for homogeneous tuples in `all` and `any`

KristofferC opened the pull request:
These were added in https://github.com/JuliaLang/julia/pull/44063 but seem to cause some ambiguities, for example in Nullables.jl:

```
  MethodError: all(::typeof(Nullables.hasvalue), ::Tuple{Nullable{Int64}, Nullable{Int64}}) is ambiguous.
  
  Candidates:
    all(f, itr::Tuple{Vararg{T, N}} where {N, T})
      @ Base reduce.jl:1298
    all(f::typeof(Nullables.hasvalue), t::Tuple)
      @ Nullables ~/.julia/packages/Nullables/RNaHb/src/nullable.jl:371
  
  Possible fix, define
    all(::typeof(Nullables.hasvalue), ::Tuple{Vararg{T, N}} where {N, T})
```

This is a pretty weird overload from Nullables but It's unclear to me if there is any benefit of having homogenous tuples fall back to the loop version.

cc @jipolanco

KristofferC committed:
remove special case for homogeneous tuples in `all` and `any`

jipolanco commented:
Ok for me. But maybe we should also ask @N5N3 who actually suggested the special case: https://github.com/JuliaLang/julia/pull/44063#issuecomment-1033305513.

I guess another solution would be to add an extra dispatch layer, something like:

```julia
any(f, itr::Tuple) = _any(f, itr)

_any(f, itr::NTuple) = _any(f, itr, :)  # case of homogeneous tuple

function _any(f, itr::Tuple)            # case of tuple with mixed types
    length(itr) > 32 && return _any(f, itr, :)
    _any_tuple(f, false, itr...)
end
```

N5N3 committed:
Use loop-based `any/all` for homogenous tuples

Unroll this loop at julia level wont gain any inference improvement, thus let LLVM unroll it if needed.

N5N3 commented:
> If there is any benefit of having homogenous tuples fall back to the loop version.

As homogenous tuples, unroll the loop at julia level wont gain any inference improvement thus I think we'd better let LLVM do this optimization for us and save some compile overhead.

Seelengrab commented:
I think this may be the same issue as #47385?

N5N3 commented:
`NTuple` is not a `Union` type so probably not?

Seelengrab commented:
Maybe :thinking: It depends on the specifics of dispatch in this case, but it at least feels similar enough :shrug:
